### PR TITLE
(PUP-5609) Protect shared state with filesystem lock

### PR DESCRIPTION
**Pull request by dalen**
Use a filesytem lock when running Puppet apply. This prevents multiple
simultaneous writes to the same state files by multiple apply processes
or a mix of apply and agent processes (the same state files are shared
by apply and agent).

This time it exits with status 1 if lock is already taken and has some tests for that from @MikaelSmith (from https://github.com/puppetlabs/puppet/pull/4582).

### CODE REVIEW

**Comment by MikaelSmith** [comment]
I'm still not sure why `tests/resource/service/puppet_mcollective_service_management.rb` fails. That needs some more investigation.

**Comment by MikaelSmith** [comment]
The acceptance test appears to run into a lag between unlinking the lockfile and exiting puppet, and when the file is released. I have to introduce a slight lag (200-300ms) between `puppet apply` runs. I want to investigate whether `puppet agent` has the same problem. This is irritating, but may not be an issue in practice.

**Comment by MikaelSmith** [comment]
My feeling is that it would be better to hold off merging until we also introduce a wait on the lockfile, and should default it to 0.5s.

**Comment by MikaelSmith** [comment]
https://github.com/puppetlabs/puppet/blob/master/acceptance/tests/agent/agent_disable_lockfile.rb covers acceptance testing the agent lockfile. It doesn't appear to test that you can do back-to-back runs.

**Comment by MikaelSmith** [comment]
@parisiale did you run into anything like this running on pxp-agent? Maybe testing service stop/start?

File: spec/unit/application/apply_spec.rb [hunk 1]
```diff
@@ -171,6 +178,30 @@
...
+    end
+
+    describe "when a filesystem lock is encountered" do
+      before :each do
+        @apply.expects(:lock).raises(Puppet::LockError, 'locked')
+        Puppet.expects(:notice).with(regexp_matches /Run of Puppet already in progress; skipping  (.+ exists)/)
+      end
+
+      it "should exit" do
+        expect { @apply.run_command }.to raise_error(SystemExit)
```
Comment (MikaelSmith) [inline spec/unit/application/apply_spec.rb]: Found out from @joshcooper that there's a helper for this, https://github.com/puppetlabs/puppet/blob/master/spec/integration/application/apply_spec.rb#L38, so we can drop lines 192-201 down to one spec test.

Comment (peterhuene) [inline spec/unit/application/apply_spec.rb]: Hi @dalen. Did you see @MikaelSmith's comment above?

**Comment by peterhuene** [comment]
Blocked on sprint triage.

**Comment by kylog** [comment]
@MikaelSmith what do you think next steps are on this one?

**Comment by MikaelSmith** [comment]
Investigate if there's something broken in the lock implementation ([PUP-3914](https://tickets.puppetlabs.com/browse/PUP-3914) comes up as a possible issue). I also left next steps at https://github.com/puppetlabs/puppet/pull/4584#issuecomment-173342757.

**Comment by hlindberg** [comment]
A thought, what happens when you debug using apply, and quit in the middle. Does that clean up and release the lock?

**Comment by MikaelSmith** [comment]
It's supposed to be cleaned up on the next run, when Puppet detects the old process is no longer running.

File: spec/unit/application/apply_spec.rb (hunk 1 above)
Comment (dalen) [inline spec/unit/application/apply_spec.rb]: Better late than never. But I've fixed that now.

**Comment by dalen** [comment]
Rebased this against master branch

**Comment by MikaelSmith** [comment]
This appears to pass acceptance tests for me, now.

**Pull request merged**